auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java

```diff
@@ -29,7 +33,7 @@
     /**
      * Roles associated with this principal.
      */
-    private final Map<String, IRole> roles;
+    private final Map<RoleKey, IRole> roles;
```

Comment:
I believe we can change this to an array or a list of IRole.

Reply:
I think we were using an array list, but as all roles where mapped to User Principal, there could be a clash with the names of roles between realm and client. This in turn could make the access control easy to by pass.
So it was decided to create `RoleKey`,`RoleType` and track `ClientId` too and store them in a Map, so we could easily save and search by the above types and params.
Also just to clarify, this went through multi changes over the last two weeks, so it could be worth syncing up @rachael-oregan to ensure best path forward to close out this PR.

Reply:
to avoid the clash, you can define the `equals` method of a role class to compare roleType, roleName and clientId. Two roles will only be the same if all three fields are the same. 

Once you have that, you can put them inside a list or array without worrying about clashing.

Reply:
If we do as I described above, I also think we don't need this many interfaces & classes to represent the roles. We could just have 1 class - UserRole. It will have 3 fields:

* type
* name
* namespace (clientId)

A realm role is just a role with a null name space. 

I think this will allow us to achieve the same object, but with far less code.

Reply:
Ah good call :+1:

Reply:
@wei-lee The main reason why a map has been used is to be able to get a role by name without the need to search through the whole list.
Since both realm and client roles can have the same name, the RoleKey class became useful to be used as key inside the map.

Reply:
@ziccardi once we have the name of the role, we can do this:

```
IRole roleToFind = new UserRole(rolename, clientId, RoleType.Client);
return roles.contains(roleToFind)
```

This way we don't need the extra classes & interfaces. I don't really see the value in them. Also I don't think it will be a problem performance wise as there won't be a lot of roles.

Reply:
@wei-lee Yes, we can, but that will loop through the whole list (in the worst case). With this implementation, we have the overhead of looping through the whole list and comparing 2 strings for equality across all the compared objects.

Reply:
Yes, I agree. But as I said, in reality, we are only looking at probably less than 10 items in the list. In this case, I don't see it become a problem. But it dramatically simplifies the implementation.